### ISSUE TITLE
Missing segments

### ISSUE THREAD

**Issue by mphilipps** [borgbackup/borg#4547]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?
Bug

## System information. For client/server mode post info for both machines.
Client:
Debian Stretch

Server:
 Debian Stretch

#### Your borg version (borg -V).
1.1.9

#### Hardware / network configuration, and filesystems used.
ssh, xfs
#### How much data is handled by borg?
From the last sucessful borg backup:

```
Duration: 1 hours 48 minutes 20.77 seconds
Number of files: 1025797
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:              490.48 GB            425.40 GB             10.85 GB
All archives:               13.81 TB             11.82 TB              1.28 TB

                       Unique chunks         Total chunks
Chunk index:                 1247169             35743323
```

#### Full borg command line that lead to the problem (leave away excludes and passwords)

prune call:
```
ionice -c 3 /etc/scripts/borg prune -s --list --remote-path /etc/scripts/borg-linux64-1.1.9 \
--show-rc --keep-within 7d --keep-daily 7 --keep-weekly 8 --keep-monthly 6 \
attic_clientservername@backupservername.domain.tld:/backup/attic_store_clientservername
```
backup call:
```
ionice -c 3 /etc/scripts/borg create -s --remote-path /etc/scripts/borg-linux64-1.1.9 \
--show-rc --exclude-caches --keep-exclude-tags **plenty of excludes here** \
attic_clientservername@backupservername.domain.tld:/backup \
/attic_store_clientservername::clientservername.domain.tld_$(date '+%Y%m%d_%H%M%S')
```
## Describe the problem you're observing.
We perform daily backups, at Tue, 2019-05-07 01:16:04 the last successfull backup happened, later that evening at 22:00 a cronjob triggered the borg prune. The borg prune failed 17 minutes later after pruning 3 archives:
Traceback (most recent call last):

```
  File "borg/remote.py", line 248, in serve

  File "borg/repository.py", line 442, in commit

  File "borg/repository.py", line 789, in compact_segments

AssertionError: Corrupted segment reference count - corrupted index or hints
```

The backup run a few hours later then revealed the following issues:
```
Remote: segment 854914 not found, but listed in compaction data
Remote: segment 908617 not found, but listed in compaction data
Remote: segment 908626 not found, but listed in compaction data
Remote: segment 908648 not found, but listed in compaction data
Remote: segment 908665 not found, but listed in compaction data
Remote: segment 908715 not found, but listed in compaction data
Remote: segment 908718 not found, but listed in compaction data
Remote: segment 908720 not found, but listed in compaction data
Remote: segment 908726 not found, but listed in compaction data
Remote: segment 908727 not found, but listed in compaction data
Remote: segment 908728 not found, but listed in compaction data
Remote: segment 908729 not found, but listed in compaction data
Remote: segment 908731 not found, but listed in compaction data
Remote: segment 922015 not found, but listed in compaction data
Remote: segment 922016 not found, but listed in compaction data
Remote: segment 922018 not found, but listed in compaction data
Remote: segment 922022 not found, but listed in compaction data
Traceback (most recent call last):

  File "borg/remote.py", line 248, in serve

  File "borg/repository.py", line 442, in commit

  File "borg/repository.py", line 789, in compact_segments

AssertionError: Corrupted segment reference count - corrupted index or hints
```
There were no recent power outages or filesystem/hardware errors.

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

I am currently running a borg check --repository-only, but it is taking a while.

Assuming I can **fix** this by running borg check with the repair flag, as in throwing out all missing segments from the compaction data, is there anyway to see which of my old backups are going to have holes in them?

**Comment by ThomasWaldmann** [comment]
If you like, you can manually have a look into your `repository/data` directory and verify that the segment files really are not there.

Before trying to "fix" this by running `borg check --repair`, you should make sure your hw / os / fs works correctly (otherwise it might get worse). Besides filesystem / os / disk issues, inconsistencies can also be caused by RAM errors (and you won't notice except if you have ECC memory and check your logs).

If you want to know for sure whether your repo / archive contents are ok, you need to run a full check, not just `--repository-only`.

**Comment by mphilipps** [comment]
I verified a couple of the missing segments, there are indeed no files to be found with that number.  
I wish I had backups of backups, because right now I have no way of telling whether these files existed at some point or were always missing. What strikes me as a bit odd is that there appear to be  3 ranges of segments missing, which assuming they are labelled in order, could correspond to the 3 pruned archives.

**Comment by ThomasWaldmann** [comment]
borg always uses last+1 for the next segment number, so every number 1 <= N <= LAST existed at some time.

some segments might be rather short-lived though as borg < 1.2 runs compaction after each repo write access and segments which are not needed any more (e.g. because their contents have been compacted into a new segment) will be deleted then.

**Comment by ThomasWaldmann** [comment]
BTW, if you have logs, you could check them if there was other noticable stuff recently except what you posted above. Esp. tracebacks, errors, warnings.

**Comment by mphilipps** [comment]
In that case it should be safe to assume that the missing segment 922022 is from april 30. The regular backup of april 30 failed, with plenty of stat: no such file for files in /var/lib/os-prober/mount/.

Following that, a manual backup was triggered (unfortunately I have no logs for that one) which succeed. 

If for whatever reason a backup fails, what happens to segments written during that backup run?

**Comment by mphilipps** [comment]
segment 908731 was from april 23, the backup on that day was successful. So that is not the culprit, or at least not  the sole culprit.

**Comment by mphilipps** [comment]
The` --repository-only` has finished with the following message.

Remote: Index mismatch for key b'\x16Sql\xba_\xeb\xc4/\x0cvC\xf9jA\xe6\x8b:\xd4\xbd\x9d\xb2\x13\xbdP\x87\xc0Wj)\x86\xb0'. (922046, 1374509) != (-1, -1)

What is the meaning of this error message?

**Comment by ThomasWaldmann** [comment]
If a borg backup run does not reach it's normal end (which commits the changes to the repo), all segments written are in uncommitted state and the repo will be rolled back to the last previous committed state (== deleting the uncommitted segment files).

That only applies to abrupt termination of borg (e.g. if borg crashes, power loss, network connectivity loss).
When borg emits a warning like "file not found" or "file not readable", that is only a warning and the backup will continue with the next file.

 It is normal for borg delete and prune to write segment files:
- first it writes (== appends, writes new segment files) a lot of DELETE objects to the repo (to mark the objects it wants to get rid of as logically deleted)
- then it commits (== write/append a COMMIT object to repo)
- then it compacts segments (segments containing PUT objects that are later deleted by DELETE objects (and are thus not needed any more) might get re-written to new compact segment files, if the old segment file has a big enough (hole size / segment size) ratio.
- then it commits again

In short: a borg repo is an implementation of a transactional log-like key/value storage. "log" here means that stuff is always appended at the end, previous stuff is never modified in-place (just deleted, if not needed any more).

The "index mismatch" means that the freshly computed repo index (by borg check, from current repo segment files' content) does not match the on-disk repo index. left side: rebuilt, right side: on-disk.

So in your case the rebuilt index has that entry, while the on-disk index does not have it.

**Comment by mphilipps** [comment]
Ok, so just because it called the failure handler, doesn't mean the run actually failed and new segments might have been committed. 
Since there is definitely something fishy going on with the repo, we decided to create a new one and pray we don't run into this issue again.

**Comment by ThomasWaldmann** [comment]
when borg "blows up" with a traceback, it won't come to commit (and we don't want to commit in that case, so that's ok). if there is no commit, it will roll back to a valid state.

**Comment by ThomasWaldmann** [comment]
Guess this is solved.

I didn't see an indication of a borg issue, rather that some files got lost from the repo (maybe hw or fs or other issue).
